File: python-vsc/problems/189.旋转数组.py
#
# @lc app=leetcode.cn id=189 lang=python3
#
# [189] 旋转数组
#

# @lc code=start
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        if not nums:
            return 
        nums_len = len(nums)
        if nums_len < 2:
            return 
        k = k % nums_len
        if not k:
            return 

        self.rotate_aux(nums, 0, nums_len - k, nums_len)



    def _gcd(self, m: int, n: int):
        while m != 0:
            t = n % m
            n = m 
            m = t
        return n

# @lc code=end






# class Solution:
#     def rotate(self, nums: List[int], k: int) -> None:
#         """
#         Do not return anything, modify nums in-place instead.
#         """
#         if not nums:
#             return 
#         nums_len = len(nums)
#         if nums_len < 2:
#             return 
#         k = k % nums_len
#         if not k:
#             return 
#         self.reserve(nums, 0, nums_len - k)
#         self.reserve(nums, nums_len - k, nums_len)
#         self.reserve(nums, 0, nums_len)

#     def reserve(self, nums: List[int], first: int, end: int):
#         last = end - 1
#         while first < last:
#             nums[first], nums[last] = nums[last], nums[first]
#             first += 1
#             last -= 1





# class Solution:
#     def rotate(self, nums: List[int], k: int) -> None:
#         """
#         Do not return anything, modify nums in-place instead.
#         """
#         if not nums:
#             return 
#         nums_len = len(nums)
#         if nums_len < 2:
#             return 
#         k = k % nums_len
#         if not k:
#             return 

#         i, j = 0, nums_len - k
#         middle = nums_len - k
#         while True:
#             while i < middle and j < nums_len:
#                 nums[i], nums[j] = nums[j], nums[i]
#                 i += 1
#                 j += 1
#             if i == middle:
#                 if j == nums_len:
#                     return
#                 middle = j
#             else:
#                 j = middle




# class Solution:
#     def rotate(self, nums: List[int], k: int) -> None:
#         """
#         Do not return anything, modify nums in-place instead.
#         """
#         if not nums:
#             return 
#         nums_len = len(nums)
#         if nums_len < 2:
#             return 
#         k = k % nums_len
#         if not k:
#             return 

#         self.rotate_aux(nums, 0, nums_len - k, nums_len)

#     def rotate_aux(self, nums: List[int], first: int, middle: int, last: int):
#         if last - first <= 1:
#             return
#         i, j = first, middle;
#         while i < middle and j < last:
#                 nums[i], nums[j] = nums[j], nums[i]
#                 i += 1
#                 j += 1
#         if i == middle:
#             if j == last:
#                 return
#             self.rotate_aux(nums, i, j, last)
#         else:
#             self.rotate_aux(nums, i, middle, last)

File: java-vsc/problems-0001/189.旋转数组.java
/*
 * @lc app=leetcode.cn id=189 lang=java
 *
 * [189] 旋转数组
 */

// @lc code=start
public class Solution {
    public void rotate(int[] nums, int k) {
        int len = nums.length;
        for (int i = 0;i < gcd(k, len);i++){
            int idx = (k + i) % len, num = nums[i];
            while (idx != i){
                int item = nums[idx];
                nums[idx] = num;
                idx = (idx + k) % len;
                num = item;
            }
            nums[i] = num;
        }
    }
    
    private int gcd(int a, int b) {
        return a % b == 0 ? b : gcd(b % a, a);
    }
}




// @lc code=end



// public class Solution {
//     public void rotate(int[] nums, int k) {
//         k = k % nums.length;
//         int count = 0;
//         for (int start = 0; count < nums.length; start++) {
//             int current = start;
//             int prev = nums[start];
//             do {
//                 int next = (current + k) % nums.length;
//                 int temp = nums[next];
//                 nums[next] = prev;
//                 prev = temp;
//                 current = next;
//                 count++;
//             } while (start != current);
//         }
//     }
// }



// public class Solution {
//     public void rotate(int[] nums, int k) {
//         k %= nums.length;
//         reverse(nums, 0, nums.length - 1);
//         reverse(nums, 0, k - 1);
//         reverse(nums, k, nums.length - 1);
//     }
//     public void reverse(int[] nums, int start, int end) {
//         while (start < end) {
//             int temp = nums[start];
//             nums[start] = nums[end];
//             nums[end] = temp;
//             start++;
//             end--;
//         }
//     }
// }